File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/CountList.java

import java.io.*;
import java.lang.*;
import java.math.*;

class CountList {
    
    /*@ public normal_behavior
      @ requires inputArray != null && (\forall int k; 0 <= k < inputArray.length; inputArray[k] != null);
      @ assignable \nothing;
      @ ensures \result == (\num_of int k; 0 <= k < inputArray.length; inputArray[k].length > 0);
      @ also
      @ public exceptional_behavior
      @ requires inputArray == null || (\exists int k; 0 <= k < inputArray.length; inputArray[k] == null);
      @ assignable \nothing;
      @ signals_only NullPointerException;
      @*/
    //@ code_java_math
    public static int countList(int[][] inputArray) {
        int count = 0;
        /*@ loop_invariant 0 <= \count <= inputArray.length;
          @ loop_invariant count == (\num_of int k; 0 <= k < \count; inputArray[k].length > 0);
          @ loop_writes count;
          @ decreases inputArray.length - \count;
          @*/
        for (int[] array : inputArray) {
            if (array.length > 0) {
                count++;
            }
        }
        return count;
    }
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/CountList.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  logic integer fb_count_rows{L}(JIntArray2 a, integer k) =
    k <= 0 ? 0 : fb_count_rows(a, k - 1) + (a->data[k - 1]->length > 0 ? 1 : 0);
*/
/*@
  requires jintarray2_valid_nonnull(inputArray);
  requires \forall integer k; 0 <= k < inputArray->length ==> inputArray->data[k] != \null;
  assigns \nothing;
  ensures \result == fb_count_rows(inputArray, inputArray->length);
*/
int32_t countList(JIntArray2 inputArray) {
    int32_t count = 0;
    int32_t length = jarray2_length(inputArray);
    /*@
      loop invariant 0 <= i <= length && length == inputArray->length;
      loop invariant count == fb_count_rows(inputArray, i) && 0 <= count <= i;
      loop assigns i, count;
      loop variant length - i;
    */
    for (int32_t i = 0; i < length; i = java_add(i, 1)) {
        JIntArray array = jarray2_get(inputArray, i);
        if (jarray_length(array) > 0) {
            count = java_add(count, 1);
        }
    }
    return count;
}
